presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/TimestampUtil.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.clickhouse;
+
+public class TimestampUtil
+{
+    private TimestampUtil() {}
+
+    public static int getMillisecondsFromTimestampString(String timestampString)
+    {
+        int milliseconds = 0;
+        int dotIndex = timestampString.indexOf('.');
+
+        if (dotIndex != -1) {
+            String fractionalPart = timestampString.substring(dotIndex + 1);
+            int fractionalLength = fractionalPart.length();
+            // Normalize the fractional part to at least 3 digits by padding with zeros. e.g., 7 -> 700; 70 -> 700
+            if (fractionalLength == 1) {
+                fractionalPart += "00";
+            }
+            else if (fractionalLength == 2) {
+                fractionalPart += "0";
+            }
```

Comment:
Do you think we should consider cases of 4, 5, 6,  etc decimal places? Also, rather than parsing a larger string, have you considered just parsing the value directly then normalizing at the end? Conditionals  are expensive and should ideally be avoided in the datatpath. It may be cheaper than the current solution, but you might need to write a microbenchmark.

Here's an example function that might get what you're going for:

https://www.desmos.com/calculator/2hheawsmd4

In case the link ever goes bad, this is the function:

$f\left(x\right)=\text{round}\left(x\cdot10^{\left(-\left(\text{floor}\left(\log10\left(x\right)\right)\ -\ 3\right)\right)}\right)$

Reply:
No because more than 3 get interpreted to seconds (as they should).
If clickhouse is storing 2025-01-08 12:34:56.7591 then the return when you do a select will be 2025-01-08 12:35:03.591 (note the 7 more seconds).

I will take a look into the rest of what you mentioned in a bit.

Reply:
Wouldn't additional values after the decimal point refer to higher precision, not necessarily more seconds? I guess if we hard code 3-decimal place precision in the clickhouse type we wouldn't expect to see more decimal places though.

Reply:
The 1st decimal place in 4 digits of precision gets translated to seconds when it gets to Presto. We hardcode the 3 decimal places when we insert from Presto. But if people do that outside of Presto, then it might be possible to have more. It's unlikely, but the common scenarios are: 3 digits for milliseconds, 6 for microseconds, and 9 for nanoseconds. The precision values can range from 0 to 9. But in case people hardcode 1 or 2, then we need the padding on presto side, otherwise .7 gets interpreted as 007 instead of 700, and .71 as 071 instead of 710

Reply:
When you say 'Also, rather than parsing a larger string', are you suggesting parsing the string decimal places directly and adding the number afterwards? Like, if it is 1 digit, then we multiply it by 100, 2 digits by 10.

Reply:
Yes because in your original code there are conditionals and allocations occurring just to create the string which is then parsed by `Integer.parseInt`. You could change the code to use mathematical operations instead to avoid allocations and conditionals.

```java
int nonNormalized = Integer.parseInt(fraction)
int normalized = (int) Math.round(fraction * Math.pow(10, -(Math.floor(Math.log10(fraction) - 3)))
return normalized;
```

Reply:
I must be missing something because desmos is off by a factor of 10 when I do the same manually. For instance, if the fraction is 75, then log10(75) is 1.875. Floored is 1. 10 to the power of 2 is 100, and multiplying that by 75 is 7500 instead of the 750 shown in desmos.

Reply:
I think what I mentioned before regarding the 4 digits is incorrect as you pointed out (ref https://clickhouse.com/docs/en/sql-reference/data-types/datetime64). It wouldn't really make sense for the digit to suddenly mean seconds. Precision refers to seconds, but it goes lower and lower